### PR TITLE
Create a `expand_matrix` method for Sparse matrices

### PR DESCRIPTION
**Pull request by Jaybsoni**
**Context:**
To improve efficiency of Op-Arithmetic we want to support sparse matrix representations of these complex arithmetic operations (Prod, Sum). This requires the need to expand the sparse matrix representation of operators to the full Hilbert space on which they are being summed or multiplied. 

**Description of the Change:**
Implement a `sparse_matrix_expand` method.

### CODE REVIEW

**Comment by Jaybsoni** [comment]
[sc-25206]

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:244]: It seems that these two cases are correctly handled by the return statement below. Is there any reason to keep this?

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:266]: These use almost the same swap matrices as the previous for loop. Can we reuse them to avoid computing them twice?

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:280]: `format` keyword missing.

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:276]: Specify that `base_matrix` should be a `scipy` sparse matrix.

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:315]: This is more Pythonic
```suggestion
                expanded_wires[i], expanded_wires[j] = expanded_wires[j], expanded_wires[i]
```

**Review by eddddddy — COMMENTED** [review]
Hey @Jaybsoni, this is looking very nice 🔥 🔥 🔥 

Just a few minor comments. I would be happy to approve this once those are addressed 🙂 .

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py:244]: Nice catch, actually when I was writing the helper functions I didn't know if `eye(0)` would behave well so I add those two edge cases in preemptively.

Comment (Jaybsoni) [inline pennylane/operation.py:244]:
```suggestion
```

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py:266]: Absolutely we can 👍🏼

File: pennylane/operation.py
Comment (Jaybsoni) [inline pennylane/operation.py:276]:
```suggestion
        base_matrix (scipy.sparse.spmatrix): base matrix to expand
```

File: pennylane/operation.py
Comment (Jaybsoni) [inline pennylane/operation.py:280]:
```suggestion
        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
            contain additional labels
        format (str): string representing the preferred scipy sparse matrix format to cast the expanded matrix too
```

File: pennylane/operation.py
Comment (AmintorDusko) [inline pennylane/operation.py:304]:
```suggestion
    num_missing_wires = n_total_wires - n_wires
    if num_missing_wires > 0:
      expanded_matrix = kron(
          base_matrix, eye(2**num_missing_wires, format=format), format=format
      )  # added missing wires at the end
```
This call of kron is not no-op if num_missing_wires=0. In some benchmarks I could see the complete sparse_expand_matrix operation being performed in close to half the time with the proposed change.

File: pennylane/operation.py
Comment (AmintorDusko) [inline pennylane/operation.py:286]: I think you could also check if the two sets of wires are in the same order. In this case, no operation is needed too.
I mean `wires=[0,1,2]` and `wire_order=[0,2,4]`.

File: pennylane/operation.py
Comment (eddddddy) [inline pennylane/operation.py:308]: Thinking about this a bit more, are you sure this outer `while` loop is necessary? It looks like after the `i`th iteration of the inner loop, all of `expanded_wires[0..i]` are now in the correct position.

Of course, I'd concede this if you find some counterexample.

**Review by eddddddy — APPROVED** [review]
Great work @Jaybsoni!

I looked through the tests in more detail and they all seem good. I just have one last question, and my approval is conditioned on it being addressed.

File: pennylane/operation.py
Comment (AmintorDusko) [reply pennylane/operation.py:286]: Sorry for this one. I forgot that wire_order must contain all wire labels.

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py:308]: I know I came across a case in the tests where not including the while loop resulted in an incorrect permutation. Let me double check and see if I can't find it again

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py:286]: I think that case should actually result in an error, its implicitly assumed that wires will always be a subset of wire_order (or exactly equal to wire_order).

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: Nice catch here 👍🏼

File: pennylane/operation.py
Comment (Jaybsoni) [inline pennylane/operation.py:303]:
```suggestion
        expanded_matrix = kron(
```

File: pennylane/operation.py
Comment (AmintorDusko) [inline pennylane/operation.py:305]:
```suggestion
    if num_missing_wires > 0:
        expanded_matrix = kron(
          base_matrix, eye(2**num_missing_wires, format=format), format=format
      )  # added missing wires at the end
    else:
      expanded_matrix = copy.copy(base_matrix)
```

File: pennylane/operation.py
Comment (AmintorDusko) [inline pennylane/operation.py:307]:
```suggestion
    U = eye(2**n_total_wires, format=format)
```

Comment (AmintorDusko) [reply pennylane/operation.py:307]: The scipy.sparse.eye defaults to the Diagonal format.

Comment (AmintorDusko) [reply pennylane/operation.py:307]: The CSR format performs better here.

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py:317]: This caused an error last time as `U.T` becomes a "csc" type sparse matrix instead of a transformed csr matrix

Comment (AmintorDusko) [reply pennylane/operation.py:319]: You are right. Thank you for the clarification.

Comment (AmintorDusko) [reply pennylane/operation.py:214]: Just to double check, did you get an error in the function?
Or the final format of the expanded_matrix was `csc` instead of `csr`?

**Review by AlbertMitjans — COMMENTED** [review]
Great job! 🚀 

Added some comments.

File: pennylane/math/matrix_manipulation.py
Comment (AmintorDusko) [inline pennylane/math/matrix_manipulation.py:212]:
```suggestion
    expanded_matrix = (U.T).asformat(format) @ expanded_matrix @ U
```

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: Consider the case `expanded_wires = [1, 2, 3, 0]` and `wire_order = [0, 1, 2, 3]`. The first index check will swap 1 <--> 2. The second index check will pass. The third index check will swap 3 <--> 0. Resulting in `expanded_wires = [2, 1, 0, 3]`

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: The type is different for the final result.

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: It breaks the for loop, but not the surrounding while loop. This is because after we do some swap, we may have caused an inconsistency at an index < current index. Since the for loop only steps forward, we need to reset it to check from the beginning again 👍🏼

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: yes nice catch!

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: This is a good question 🤔, I know that the existing expand method keeps `wire_order` as a key word argument. It makes sense that it should be a mandatory argument otherwise what would be the purpose of such a function? Might be worth discussing if this is a change we can propagate with out too much refactoring in the code.

File: pennylane/math/matrix_manipulation.py
Comment (Jaybsoni) [reply pennylane/math/matrix_manipulation.py]: Added, but I casted the final expanded matrix 👍🏼

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: It was complaining earlier, but this line might have changed since then. Let me double check

File: pennylane/operation.py
Comment (eddddddy) [reply pennylane/operation.py]: Right, how about instead of
```python
j = wire_order.index(expanded_wires[i])
```
we have
```python
j = expanded_wires.index(wire_order[i])
```
and remove the `break` statement. I think this way we maintain the loop invariant that `expanded_wires[0:i + 1] == wire_order[0:i + 1]` after iteration `i`.

File: pennylane/math/matrix_manipulation.py
Comment (AmintorDusko) [reply pennylane/math/matrix_manipulation.py]: Yeah. That's why I asked if you were really getting an error or the wrong sparse format returned by the function.

File: pennylane/operation.py
Comment (AmintorDusko) [reply pennylane/operation.py]: Sound good.

**Review by AmintorDusko — APPROVED** [review]
Nice work! Thanks for that!

File: tests/math/test_matrix_manipulation.py
Comment (Jaybsoni) [inline tests/math/test_matrix_manipulation.py:521]: duplicate
```suggestion
```

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: Nice! I think this should work in all cases!

Comment (Jaybsoni) [reply pennylane/operation.py]: Added 👍🏼

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: This has since been removed in favour of a slightly different yet more performant swapping algorithm

File: pennylane/operation.py
Comment (Jaybsoni) [reply pennylane/operation.py]: For now we stick to the default implementation of `expand_matrix()`

**Pull request merged**
